x-pack/plugins/fleet/server/services/spaces/enable_space_awareness.ts

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { SavedObjectsClientContract } from '@kbn/core-saved-objects-api-server';
+
+import { LEGACY_AGENT_POLICY_SAVED_OBJECT_TYPE } from '../../../common';
+import {
+  LEGACY_PACKAGE_POLICY_SAVED_OBJECT_TYPE,
+  PACKAGE_POLICY_SAVED_OBJECT_TYPE,
+} from '../../../common/constants';
+
+import { appContextService } from '..';
+import { AGENT_POLICY_SAVED_OBJECT_TYPE } from '../../constants';
+import { getSettings, saveSettings } from '../settings';
+import { FleetError } from '../../errors';
+
+const PENDING_TIMEOUT = 60 * 60 * 1000;
+
+export async function enableSpaceAwarenessMigration() {
+  const soClient = appContextService.getInternalUserSOClientWithoutSpaceExtension();
+
+  const existingSettings = await getSettings(soClient);
+  if (existingSettings.use_space_awareness) {
+    throw new FleetError('Migration is already done.');
+  }
+
+  if (
+    existingSettings.use_space_awareness_migration_started_at &&
+    new Date(existingSettings.use_space_awareness_migration_started_at).getTime() >
+      Date.now() - PENDING_TIMEOUT
+  ) {
+    throw new FleetError('Migration is pending.');
+  }
+
+  await saveSettings(
+    soClient,
+    {
+      use_space_awareness_migration_started_at: new Date().toISOString(),
+    },
+    {
+      version: existingSettings.version,
+    }
+  );
+
+  // Migration
+  // Agent Policy
+  await batchMigration(
+    soClient,
+    LEGACY_AGENT_POLICY_SAVED_OBJECT_TYPE,
+    AGENT_POLICY_SAVED_OBJECT_TYPE
+  );
+  await batchMigration(
+    soClient,
+    LEGACY_PACKAGE_POLICY_SAVED_OBJECT_TYPE,
+    PACKAGE_POLICY_SAVED_OBJECT_TYPE
+  );
+
+  // Todo
+  // uninstall token
+  // message signing
+
+  // Update Settings SO
+  await saveSettings(soClient, {
+    use_space_awareness: true,
+    use_space_awareness_migration_started_at: null,
+  });
+}
+
+const BATCH_SIZE = 1000;
+
+async function batchMigration(
+  soClient: SavedObjectsClientContract,
+  sourceSoType: string,
+  targetSoType: string
+) {
+  const finder = soClient.createPointInTimeFinder({
+    type: sourceSoType,
+    perPage: BATCH_SIZE,
+  });
+  try {
+    for await (const result of finder.find()) {
+      const createRes = await soClient.bulkCreate<any>(
```

Comment:
Are there objects referencing this "moving" objects that need to be updated too?

Reply:
We do not use references for Fleet saved objects (probably something we can improve in the future as there is relations between our object)